docs: add complex payload schema in AsyncAPI Document (#1889)
asyncapi/website: pages/docs/concepts/asyncapi-document/payload-schema.md

---
title: Payload Schema

> weight: 239

Reviewer: Either this page weight number is not sufficiently high, or the weight in the `_section.md` needs to be fixed... do you notice in your preview link that the `Overview` page shows incorrectly AFTER this new document?

BTW, moving forward, let's remember to always review your work via the generated Netlify preview link generated in each PR. Ideally, you should have noticed this before I did 😄

<img width="300" alt="Screen Shot 2023-08-31 at 6 20 29 PM" src="https://github.com/asyncapi/website/assets/19964402/5932e56b-57f8-40c3-a0d4-63eb2c273bad">

---

The payload schema defines a message's format, data types, and properties to ensure that the payload follows a specific structure and data format.

> In the AsyncAPI specification, you can use schema languages such as JSON or Avro Schema to create the payload schema. Enabling consumers to gain insights into the structure and data types of the payload.

Reviewer 2: ```suggestion
In the AsyncAPI specification, you can use schema languages such as JSON Schema or Avro to create the payload schema. Enabling consumers to gain insights into the structure and data types of the payload.
```

Do not split JSON and Schema because the name is JSON Schema. [JSON](https://www.json.org/json-en.html) and [JSON Schema](https://json-schema.org/) are two different things.

> You should use Avro schemas to handle complex messages and structured data models. Avro schemas support complex types like arrays and records. You can create Avro schemas by writing the JSON representation or using tools like Avro4s UI to generate the schema from valid JSON.

Reviewer 2: Why recommend Avro for complex schemas? Actually, JSON Schema is more suitable for complex structures. I would be careful of recommending one over the other because it depends on multiple factors.

Reviewer: Based on this feedback, it leads me to think we need to add a section for "Define JSON schema". (Is this accurate @fmvilas/ @derberg?

Reviewer 3: So basically there are multiple schemas, in document that @kakabisht writes, I recommend to focus on default `AsyncAPI Schema` (that is basically the same - superset - as JSON Schema but with few additional fields), then you can have separate documents (I think easier to find, easier to digest) for JSON Schema, Avro, Protobuf

Reviewer 2: Yes, but I'd still avoid "recommending" one over the other. I know it's been changed but mentioning it again for clarity.

Reviewer 3: yes, no recommendations, just documenting what is possible, and what are the limitations

Author: Thank you for the feedback.

> You should use Avro schemas to handle complex messages and structured data models. Avro schemas support complex types like arrays and records. You can create Avro schemas by writing the JSON representation or using tools like Avro4s UI to generate the schema from valid JSON.

Reviewer: You should link directly to the tool you mentioned, Avro4s UI.

Use npm or yarn to install the `@asyncapi/avro-schema-parser` package to enable Avro schema parsing in AsyncAPI.

## Define Avro schema

> Define the Avro schema for the message payload. You can choose one of the following methods,

Reviewer 2: ```suggestion
Define the Avro schema for the message payload. You can choose one of the following methods:
```

- Embedded notation: Define the Avro schema within the message payload property.
- Remote reference: Specify the schema using an absolute remote endpoint, such as `$ref: 'https://schemas.example.com/user'`.

> - Local reference: Specify the schema using a relative reference, such as `$ref: './user-signedup.avsc#/User'`.

Reviewer 2: ```suggestion
- Local file reference: Specify the schema using a relative reference, such as `$ref: './user-signedup.avsc'`.
```

The diagram below defines the AsyncAPI specification file using the local reference method.

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;
style E fill:#47BCEE, stroke:#333, stroke-width:2px;

A[Define Avro Schema]
B[Local Reference]
C[Avro Schema File]
D[Message Payload]
E[AsyncAPI Specification File]

A -->|One of the following methods:| B
B -->|Specify schema using relative reference| D
E -->|Uses local reference method| C
C-->|Defines structure of message payload| D

> ```

Reviewer 2: Not sure I understand the purpose of this diagram 🤔

Reviewer: As mentioned in another PRs, I've explained to Fran in our Spec 3 Docs meeting why we are adding diagrams in our project scope. Fran should now understand why we have several diagrams.

Reviewer 3: I understand we need diagrams, but not sure what this diagram represents. 

> The diagram below defines the AsyncAPI specification file using the local reference method.

I don't think diagram `defines the AsyncAPI specification file using the local reference method`

Reviewer 2: Yeah, I find it confusing. Still not sure of the meaning of this diagram, even after you explained it @alequetzalli.

Author: @derberg @fmvilas , i updated the diagram. Could you help me understand how i can make this image better?

Reviewer 3: so diagram description says:
`The diagram below defines the AsyncAPI specification file using the local reference method.`

I find it confusing as I cannot imagine how diagram can define something, and how `local` is special in the scenario? btw there is also a way to specify Avro in `components` and also have a reference to it from `payload.schema`

I basically suggest you to look on the diagram from my perspective, a doc reader. Diagrams are like text, we "read" them, component by component, and try to understand that they mean. So try to read it, and explain with text what it shows cause I do not even know what it means that `Define Avro Schema` -> `specifies`? like who specifies? What is the goal of your diagram

As example look at https://www.asyncapi.com/docs/tools/generator#generation-process where you see an example, where diagram supplements a detailed description of generation process, where text match diagram components names. Similar to https://www.asyncapi.com/docs/tools/generator/asyncapi-document#asyncapi-document-generation-process

> Here is an example of an AsyncAPI specification file that uses the local reference method,

Reviewer 2: ```suggestion
Here is an example of an AsyncAPI specification file that uses the local reference method:
```

```yaml
messageId: userSignup
name: UserSignup
title: User signup
summary: Action to sign a user up.
description: A longer description
tags:
  - name: user
  - name: signup
payload:
  schemaFormat: application/vnd.apache.avro+json;version=1.9.0
  schema:

>     $ref: path/to/user-create.avsc#/UserCreate

Reviewer 2: ```suggestion
    $ref: path/to/user-create.avsc
```

```

> Create a separate Avro schema file with a .avsc extension. The file should define the structure of the message payload. Here is an example of an Avro schema file for the User record type:

Reviewer: ```suggestion
Create a separate Avro schema file with a `.avsc` extension. The file should define the structure of the message payload. Here is an example of an Avro schema file for the user record type:
```

```yaml
namespace: UserCreate.avro
type: record
name: user
fields:
  - name: fullName
    type: string
  - name: email
    type: string

> ```

Reviewer 2: AVSC files should be JSON instead of YAML. It may work like this because our parser does the conversion on the fly but this is an invalid AVSC file.

## Attach examples

Although optional, it is highly recommended to attach examples to the AsyncAPI specification. You can use JSON or YAML format for binary encodings, like Avro. Attach the examples to the examples property within the message payload definition. Here is an example,

```yaml
examples:
  - name: SimpleSignup
    summary: A simple UserSignup example message
  - payload:
      user:
        fullName: Demo
        email: demo@demo.io
```

You can use a Schema Registry to separate the Avro schema from the message payload, making it easier to manage schema compatibility.

## Reuse schema

> To reuse a schema in your AsyncAPI specification, define it in the components/schemas section and reference it using the `$ref` keyword. Using `$ref` helps to avoid duplication and ensures consistency. Here's an example of reusing a schema from components in AsyncAPI.

Reviewer: ```suggestion
To reuse a schema in your AsyncAPI specification, define it in the `components/schemas` section and reference it using the `$ref` keyword. Using `$ref` helps to avoid duplication and ensures consistency. Here's an example of reusing a schema from components in AsyncAPI.
```
